src/V3EmitMk.cpp

```diff
@@ -24,18 +24,514 @@
+        bool m_isConcatenable = true;
+        // Work list ID for debugging.
+        const int m_dbgId;
+
+        WorkList() = delete;
+        WorkList(int id)
+            : m_dbgId{id} {}
+    };
+
+    // MAIN PARAMETERS
+
+    // Minimum number of input files required to perform concatenation.
+    // Concatenation of a small number of files does not give any performance advantages.
+    // The value has been chosen arbitrarily, most likely could be larger.
+    static constexpr size_t MIN_FILES_COUNT = 100;
+
+    // Concatenation of only a few files most likely does not increase performance.
+    // The value has been chosen arbitrarily.
+    static constexpr size_t MIN_FILES_PER_BUCKET = 4;
+
+    // MEMBERS
+
+    std::vector<FilenameWithScore> m_inputFiles;
+    std::vector<FileOrConcatenatedFilesList> m_outputFiles;
+    uint64_t m_totalScore;
+    std::string m_groupFilePrefix;
+    std::vector<WorkList> m_workLists;
+
+    OutputGroup(std::vector<FilenameWithScore> inputFiles, uint64_t totalScore,
+                std::string groupFilePrefix)
+        : m_inputFiles{std::move(inputFiles)}
+        , m_totalScore{totalScore}
+        , m_groupFilePrefix{groupFilePrefix} {}
+
+    // Debug logging: prints scores histogram
+    static void dumpLogScoreHistogram(std::ostream& os,
+                                      const std::vector<uint64_t>& sortedScores) {
+        constexpr int MAX_BAR_LENGTH = 80;
+        constexpr int MAX_INTERVALS_NUM = 60;
+
+        const int64_t topScore = sortedScores.back();
+        os << "Top score: " << topScore << endl;
+        const int maxScoreWidth = std::to_string(topScore).length();
+
+        const int64_t intervalsNum = std::min<int64_t>(topScore + 1, MAX_INTERVALS_NUM);
+
+        struct Interval final {
+            uint64_t m_lowerBound = 0;
+            int m_size = 0;
+        };
+
+        std::vector<Interval> intervals;
+        intervals.resize(intervalsNum);
+
+        for (uint64_t score = topScore; score >= 1; --score) {
+            const unsigned int ivIdx = score * intervalsNum / (topScore + 1);
+            intervals[ivIdx].m_lowerBound = score;
+        }
+        intervals[0].m_lowerBound = 0;
+
+        for (const auto score : sortedScores) {
+            const unsigned int ivIdx = score * intervalsNum / (topScore + 1);
+            ++intervals[ivIdx].m_size;
+        }
+
+        int topIntervalSize = 0;
+        for (const auto& iv : intervals) topIntervalSize = std::max(topIntervalSize, iv.m_size);
+
+        os << "Input files' scores histogram:" << endl;
+
+        for (int ivIdx = 0; ivIdx < intervalsNum; ++ivIdx) {
+            const Interval& iv = intervals[ivIdx];
+            const int scaledSize = iv.m_size * (MAX_BAR_LENGTH + 1) / topIntervalSize;
+
+            std::string line = " |" + std::string(scaledSize, '#');
+
+            os << std::setw(maxScoreWidth) << iv.m_lowerBound << line << "  " << iv.m_size << endl;
```

Comment:
```suggestion
            os << std::setw(maxScoreWidth) << iv.m_lowerBound << line << "  " << iv.m_size << "\n";
```

Avoids frequent flush